train.py

```diff
@@ -232,32 +237,42 @@ def train_attack(config):
         # with create_profiler(enabled=config.get('profile', False)) as prof:
         for batch in tqdm(train_loader):
             batch = {k: v.to(config['device']) for k, v in batch.items()}
-
-            try:
-                with record_function('training_step'):
-                    attack.train()
-                    loss = attack.step(batch)
-                    loss.backward()
-
-                    cur_lr = scheduler.get_last_lr()[0]
-                    log_info({'train/loss': loss, 'train/lr': cur_lr}, step=step)
-                    
-                    attack.pre_update(optim)
-                    optim.step()
-                    scheduler.step()
-                    optim.zero_grad()
-                    attack.post_update(optim)
-
-            except RuntimeError as e:
-                logger.info(f'encountered an error at step={step}:\n{e}')
+            if config.get('attack_type') == 'fgsm':
```

Comment:
one idea for fgsm is rather than doing the one-step attack, we could do a similar iterative update as detailed here: https://www.neuralception.com/adversarialexamples-bim/

Reply:
i'm also now realizing that most of these perturbation attacks learn image-_dependent_ attacks. presumably learning universal perturbations would work but be less accurate. this does give some nice contrast to the patch-based stuff though because those are specifically meant to be image independent.

Reply:
i didn't realize this until recently so apologies for writing all the boilerplate in a way that assumed all methods were universal attacks

Reply:
@tjbai also something i have been thinking about -universal attack pretty challenging.  I am going to try fixing image class and see how that goes